File: be/join.py
#!/usr/bin/env python3
"""
be join — onboard to BountyNet via Dynamic (GitHub login → wallet → agent ID).

Usage:
  be join

Flow:
  1. Opens browser → Dynamic auth (GitHub OAuth / email / wallet)
  2. Dynamic creates embedded wallet
  3. Gateway registers agent on EIP-8004 Identity Registry (Arc)
  4. CLI stores credentials locally

Result: ~/.bountynet/agent.json
  {
    "agent_id": 1,
    "wallet": "0x...",
    "ens": "agent-1.maceip.eth",
    "token": "dyn_..."
  }
"""
import os
import sys
import json
import time
import http.server
import threading
import webbrowser

GATEWAY = os.environ.get("BOUNTYNET_GATEWAY", "https://gateway.stare.network")
DYNAMIC_ENV_ID = os.environ.get("DYNAMIC_ENV_ID", "36a24240-ece2-4568-a45e-463437650d21")
CONFIG_DIR = os.path.expanduser("~/.bountynet")
CONFIG_FILE = os.path.join(CONFIG_DIR, "agent.json")

# Dynamic hosted auth URL — redirects back to localhost after login
AUTH_URL = f"https://app.dynamic.xyz/connect/{DYNAMIC_ENV_ID}"
CALLBACK_PORT = 9876


def log(msg):
    sys.stderr.write(f"[be] {msg}\n")
    sys.stderr.flush()


def existing_config():
    if os.path.exists(CONFIG_FILE):
        with open(CONFIG_FILE) as f:
            return json.load(f)
    return None


def save_config(data):
    os.makedirs(CONFIG_DIR, exist_ok=True)
    with open(CONFIG_FILE, "w") as f:
        json.dump(data, f, indent=2)
    os.chmod(CONFIG_FILE, 0o600)


def onboard_via_gateway(dynamic_token: str) -> dict:
    """Call gateway to register agent on-chain."""
    import requests
    resp = requests.post(f"{GATEWAY}/identity/onboard", json={
        "dynamic_token": dynamic_token,
    }, headers={
        "Authorization": f"Bearer {dynamic_token}",
    }, timeout=30)
    return resp.json()


class CallbackHandler(http.server.BaseHTTPRequestHandler):
    """Receives the OAuth callback from Dynamic."""
    token = None

    def do_GET(self):
        from urllib.parse import urlparse, parse_qs
        query = parse_qs(urlparse(self.path).query)
        token = query.get("token", [None])[0] or query.get("jwt", [None])[0]

        if token:
            CallbackHandler.token = token
            self.send_response(200)
            self.send_header("Content-Type", "text/html")
            self.end_headers()
            self.wfile.write(b"""
                <html><body style="font-family:system-ui;text-align:center;padding:4em">
                <h1>BountyNet</h1>
                <p>Authenticated. You can close this tab.</p>
                </body></html>
            """)
        else:
            self.send_response(400)
            self.end_headers()
            self.wfile.write(b"Missing token")

    def log_message(self, *args):
        pass


def wait_for_callback(timeout=120):
    """Start local server, wait for Dynamic callback."""
    server = http.server.HTTPServer(("127.0.0.1", CALLBACK_PORT), CallbackHandler)
    server.timeout = timeout

    def serve():
        while CallbackHandler.token is None:
            server.handle_request()

    t = threading.Thread(target=serve, daemon=True)
    t.start()

    # Open browser
    callback_url = f"http://localhost:{CALLBACK_PORT}/callback"
    auth_url = f"{AUTH_URL}?redirect_uri={callback_url}"
    log(f"opening browser for login...")
    webbrowser.open(auth_url)

    # Wait
    deadline = time.time() + timeout
    while CallbackHandler.token is None and time.time() < deadline:
        time.sleep(0.5)

    server.server_close()
    return CallbackHandler.token


def main():
    # Check existing config
    existing = existing_config()
    if existing:
        log(f"already joined as agent #{existing.get('agent_id')}")
        log(f"wallet: {existing.get('wallet')}")
        log(f"ens: {existing.get('ens')}")
        log(f"to re-join, delete {CONFIG_FILE}")
        return

    log("joining BountyNet...")

    # Get Dynamic auth token
    token = wait_for_callback()
    if not token:
        log("login timed out — try again")
        sys.exit(1)

    log("authenticated — registering agent...")

    # Register via gateway
    try:
        result = onboard_via_gateway(token)
    except Exception as e:
        log(f"gateway error: {e}")
        sys.exit(1)

    if "error" in result:
        log(f"registration failed: {result['error']}")
        sys.exit(1)

    config = {
        "agent_id": result.get("agent_id"),
        "wallet": result.get("wallet"),
        "ens": result.get("ens"),
        "token": token,
        "gateway": GATEWAY,
    }

    save_config(config)

    log(f"")
    log(f"joined BountyNet!")
    log(f"  agent:  #{config['agent_id']}")
    log(f"  wallet: {config['wallet']}")
    log(f"  ens:    {config['ens']}")
    log(f"")
    log(f"next: be watch")


if __name__ == "__main__":
    main()

File: be/src/cli/join.rs
use crate::dirs;
use clap::Args;
use eyre::Result;
use serde_derive::{Deserialize, Serialize};
use std::fs;

/// Join the BountyNet network.
///
/// Opens a browser for Dynamic login (GitHub OAuth).
/// Creates a wallet and registers an agent identity on Arc.
/// Saves credentials to ~/.bountynet/agent.json.
#[derive(Debug, Args)]
pub struct Join {
    /// Gateway URL
    #[clap(long, default_value = "https://gateway.stare.network")]
    pub gateway: String,

    /// Skip browser open (print URL instead)
    #[clap(long)]
    pub no_browser: bool,
}

#[derive(Serialize, Deserialize, Debug)]
struct AgentConfig {
    agent_id: u64,
    wallet: String,
    ens: String,
    token: String,
    gateway: String,
}

#[derive(Deserialize)]
struct OnboardResponse {
    agent_id: Option<u64>,
    wallet: Option<String>,
    ens: Option<String>,
    error: Option<String>,
}

impl Join {
    pub fn run(self) -> Result<()> {
        let config_dir = dirs::HOME.join(".bountynet");
        let config_file = config_dir.join("agent.json");

        // Check existing
        if config_file.exists() {
            let existing: AgentConfig = serde_json::from_str(&fs::read_to_string(&config_file)?)?;
            eprintln!("[bounty] already joined as agent #{}", existing.agent_id);
            eprintln!("[bounty] wallet: {}", existing.wallet);
            eprintln!("[bounty] ens: {}", existing.ens);
            eprintln!("[bounty] to re-join, delete {}", config_file.display());
            return Ok(());
        }

        eprintln!("[bounty] joining BountyNet...");

        // Start local callback server
        let server = tiny_http::Server::http("127.0.0.1:9876")
            .map_err(|e| eyre::eyre!("failed to start callback server: {}", e))?;

        // Open browser to Dynamic auth
        let callback = "http://localhost:9876/callback";
        let auth_url = format!("{}/identity/login?redirect_uri={}", self.gateway, callback);

        if self.no_browser {
            eprintln!("[bounty] open this URL in your browser:");
            eprintln!("  {}", auth_url);
        } else {
            eprintln!("[bounty] opening browser for login...");
            let _ = open::that(&auth_url);
        }

        // Wait for callback with token
        eprintln!("[bounty] waiting for login...");
        let token = wait_for_callback(&server)?;
        eprintln!("[bounty] authenticated");

        // Register via gateway
        eprintln!("[bounty] registering agent...");
        let client = reqwest::blocking::Client::new();
        let resp = client
            .post(format!("{}/identity/onboard", self.gateway))
            .header("Authorization", format!("Bearer {}", token))
            .json(&serde_json::json!({
                "external_id": format!("dynamic:{}", token),
                "dynamic_token": token,
            }))
            .send()?;

        let data: OnboardResponse = resp.json()?;

        if let Some(err) = data.error {
            eprintln!("[bounty] registration failed: {}", err);
            return Err(eyre::eyre!("registration failed: {}", err));
        }

        let config = AgentConfig {
            agent_id: data.agent_id.unwrap_or(0),
            wallet: data.wallet.unwrap_or_default(),
            ens: data.ens.unwrap_or_default(),
            token: token.clone(),
            gateway: self.gateway.clone(),
        };

        // Save config
        fs::create_dir_all(&config_dir)?;
        fs::write(&config_file, serde_json::to_string_pretty(&config)?)?;

        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            fs::set_permissions(&config_file, fs::Permissions::from_mode(0o600))?;
        }

        eprintln!();
        eprintln!("[bounty] joined BountyNet!");
        eprintln!("[bounty]   agent:  #{}", config.agent_id);
        eprintln!("[bounty]   wallet: {}", config.wallet);
        eprintln!("[bounty]   ens:    {}", config.ens);
        eprintln!();
        eprintln!("[bounty] next: bounty watch");

        Ok(())
    }
}

fn wait_for_callback(server: &tiny_http::Server) -> Result<String> {
    loop {
        let request = server
            .recv()
            .map_err(|e| eyre::eyre!("server error: {}", e))?;

        let url = request.url().to_string();

        // Parse token from query string
        if let Some(query) = url.split('?').nth(1) {
            for param in query.split('&') {
                let mut parts = param.splitn(2, '=');
                let key = parts.next().unwrap_or("");
                let value = parts.next().unwrap_or("");
                if key == "token" || key == "jwt" {
                    // Respond with success page
                    let response = tiny_http::Response::from_string(
                        "<html><body style='font-family:system-ui;text-align:center;padding:4em'>\
                         <h1>BountyNet</h1>\
                         <p>Authenticated. You can close this tab.</p>\
                         </body></html>",
                    )
                    .with_header(
                        "Content-Type: text/html"
                            .parse::<tiny_http::Header>()
                            .unwrap(),
                    );
                    let _ = request.respond(response);
                    return Ok(value.to_string());
                }
            }
        }

        // Not the callback we're looking for
        let response = tiny_http::Response::from_string("waiting for auth...");
        let _ = request.respond(response);
    }
}
